### PR TITLE
Normal Service takes priority over Externalname service

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 1]
```diff
@@ -233,11 +233,9 @@ func (c *Controller) InstancesByPort(svc *model.Service, port int,
 		var err error
 		tmpInstances, err = r.InstancesByPort(svc, port, labels)
 		if err != nil {
+			log.Warnf("Instances() found match but encountered an error: %v", errs)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/aggregate/controller.go:236]: nit: log is wrong, `len(tmpInstances)` might be 0

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -678,6 +676,10 @@ func (c *Controller) WorkloadHealthCheckInfo(addr string) model.ProbeList {
 // InstancesByPort implements a service catalog operation
 func (c *Controller) InstancesByPort(svc *model.Service, reqSvcPort int,
 	labelsList labels.Collection) ([]*model.ServiceInstance, error) {
+	res, _ := c.endpoints.InstancesByPort(c, svc, reqSvcPort, labelsList)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:679]: why ignore error?

**Comment by hzxuzhonghu** [comment]
Fixed, just realize i forgot to push.

**Comment by hzxuzhonghu** [comment]
ping @howardjohn

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -321,9 +321,7 @@ func (c *Controller) onServiceEvent(curr interface{}, event model.Event) error {
 			c.nodeSelectorsForServices[svcConv.Hostname] = getNodeSelectorsForService(*svc)
 		}
 		c.servicesMap[svcConv.Hostname] = svcConv
-		if instances == nil {
-			delete(c.externalNameSvcInstanceMap, svcConv.Hostname)
-		} else {
+		if len(instances) > 0 {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:324]: why is this removed? won't we leak now?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:324]: What i did here is only add to `externalNameSvcInstanceMap` when it is a real externalName service with instances. So absolutely no leak.
